File: node/kmeans.js
var _ = require("lodash");
var fs = require("fs");

function sq(x) { return x * x }

function dist(x, y) {
  return Math.sqrt(sq(x[0] - y[0]) + sq(x[1] - y[1]));
}

function closest(x, choices) {
  return _.min(choices, function(y) { return dist(x, y); });
}

function clusters(xs, centroids) {
  return _(xs)
    .groupBy(function (x) { return closest(x, centroids); })
    .values()
    .value();
}

function vsum(v, w) {
  return _.map(v, function(x, i) { return x + w[i]; });
}

function average(xs) {
  var total = _.reduce(xs, vsum);
  var l = xs.length;

  return _.map(total, function(t) { return t / l; });
}

function run(xs, n, iters) {
  var centroids = _.take(xs, n);
  for (var i = 0; i < iters; i++) {
    centroids = _.map(clusters(xs, centroids), average);
  }
  return clusters(xs, centroids);
}


var points = JSON.parse(fs.readFileSync("../points.json"));
var iterations = 100;
var n = 10;
var start = (new Date()).getTime();
for (var i = 0; i < iterations; i++) {
  run(points, n, 15);
}
var time = ((new Date()).getTime() - start) / iterations;

console.log("Running " + iterations + " iterations as required " + time + " ms");

File: ruby/kmeans.rb
require 'json'

class Point
  attr_accessor :x
  attr_accessor :y

  def initialize(x, y)
    @x = x
    @y = y
  end

  def +(other)
    Point.new @x + other.x, @y + other.y
  end

  def /(value)
    Point.new @x / value.to_f, @y / value.to_f
  end

  def dist(other)
    Math.sqrt((@x - other.x) ** 2 + (@y - other.y) ** 2)
  end

  def closest(points)
    points.min { |a, b|
      self.dist(a) <=> self.dist(b)
    }
  end

  def inspect
    '(%.6f,%.6f)' % [@x, @y]
  end
end

def update_centroids(points, centroids)
  groups = groupby(points, centroids)
  res = []
  groups.each { |k, v|
    res << v.inject(Point.new(0,0), :+) / v.length
  }
  res
end

def groupby(points, centroids)
  g = Hash.new {|h,k| h[k]=[]}
  points.each { |pp|
    c = pp.closest(centroids)
    g[c] << pp
  }
  g
end

def run(xs, n, iters=15)
  centroids = xs.take(n)
  (1..iters).each {
    centroids = update_centroids(xs, centroids)
  }
  groupby(xs, centroids)
end

points = JSON.parse(open("../points.json").read).map { |p|
  Point.new p[0], p[1]
}

iterations = 100
start = Time.now
(1..iterations).each {
  run(points, 10)
}
total = (Time.now - start) * 1000 / iterations
puts 'Made % iterations with an average of %.6f milliseconds' % [iterations, total]
